include/envoy/formatter/substitution_formatter.h

```diff
@@ -78,5 +79,47 @@ class FormatterProvider {
 
 using FormatterProviderPtr = std::unique_ptr<FormatterProvider>;
 
+/**
+ * Interface for command parser.
+ * CommandParser returns a FormatterProviderPtr after successfully parsing an access log format
+ * token, nullptr otherwise.
+ */
+class CommandParser {
+public:
+  virtual ~CommandParser() = default;
+
+  /**
+   * Return a FormatterProviderPtr if this command is parsed from the token.
+   * @param token the token to parse
+   * @param pos current position in the entire format string
+   * @param command_end_position position at the end of the command token
+   * @return FormattterProviderPtr substitution provider for the parsed command
+   */
+  virtual FormatterProviderPtr parse(const std::string& token, size_t pos,
+                                     int command_end_position) const PURE;
```

Comment:
nit: is there any reason to use int here vs. size_t?

Reply:
not really, fixed